Adding decimal support for python client generation

m-standfuss opened the pull request:
<!-- Enter details of the change here. Include additional tests that have been done, reference to the issue for tracking, etc. -->
Adding decimal support for python client generation

This PR adds the ability to handle Decimal objects coming in and out of python servers. We "recognize" the open api `decimal` field and map it correctly to the python Decimal class. I consider this a non-breaking change because prior to this if a decimal field was used to generate python client library it just didnt work on decimal fields. 

https://github.com/OpenAPITools/openapi-generator/issues/19198

<!-- Please check the completed items below -->
### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh ./bin/configs/*.yaml
  ./bin/utils/export_docs_generators.sh
  ``` 
  (For Windows users, please run the script in [Git BASH](https://gitforwindows.org/))
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  IMPORTANT: Do **NOT** purge/delete any folders/files (e.g. tests) when regenerating the samples as manually written tests may be removed.
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (upcoming 7.6.0 minor release - breaking changes with fallbacks), `8.0.x` (breaking changes without fallbacks)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

CC Python community for :eyes: 
@cbornet @tomplus @krjakbrjak @fa0311 @multani

multani reviewed (changes requested):
When you have time, could you regenerate the sample files as explained in the pull request checklist? It should update several files automatically.

Would you also be able to add a couple of tests to verify that serializing/deserializing objects with decimal work as expected? This would be useful to prevent regression in the future. The "echo client" sample should be a good starting point.

Thanks!

m-standfuss commented:
@multani thank you for looking. I have ran the build steps now and pushed up the commit. Thank you for the example test I will get some coverage added to the branch this weekend and hit you back up to look.

m-standfuss commented:
@multani curious if you can shoot a link to the "echo client" you were referring to? I found this but just wanted to make sure I am in the right spot as there were a couple other spots that looked similar.

multani commented:
I was thinking of this sample: https://github.com/OpenAPITools/openapi-generator/tree/master/samples/client/echo_api/python

One example of a past change:
https://github.com/OpenAPITools/openapi-generator/commit/5e9546451c829738afa920c28121727f0b5c89a8

wing328 commented:
@m-standfuss here is another good starting point to add some tests:

https://github.com/OpenAPITools/openapi-generator/blob/b5d5e572e01b41305c6f49615e312f13ac6dc107/samples/openapi3/client/petstore/python/tests/test_api_client.py#L193

you can find other tests for `sanitize_for_serialization` in that test file

m-standfuss commented:
Thank you @wing328 and @multani! will get some good coverage included.

m-standfuss commented:
@wing328 and @multani sorry took a minute to get these running locally. 

 Added to an existing test for serialization and added a new test for deserialization. Please let me know if they are not in the right spot or you would like to see more. Thanks again!

multani reviewed (approved):
Thanks, LGTM :+1:

multani commented on samples/openapi3/client/petstore/python-pydantic-v1/tests/test_api_client.py:
```diff
@@ -156,6 +155,11 @@ def test_sanitize_for_serialization(self):
         result = self.api_client.sanitize_for_serialization(data)
         self.assertEqual(result, "1997-07-16T19:20:30.450000+01:00")
 
+        # decimal
+        data = Decimal("1.0")
+        result = self.api_client.sanitize_for_serialization(data)
+        self.assertEquals(result, "1.0")
+
```
Actually, can you move this into `samples/openapi3/client/petstore/python/tests/test_api_client.py` instead?

The "pydantic v1" backend will be removed at some point.

m-standfuss replied:
oh good catch, thats on me, fixing now

m-standfuss replied:
3724025e0def4b2713a85d623df7dcbb4afcfe1a

m-standfuss replied:
@multani @wing328  do i need to worry about the failing steps in the build? They appear to be failures to install poetry for the v1 petstore sample. 

```
Run pipx install --python '/opt/hostedtoolcache/Python/3.7.17/x64/bin/python' poetry
creating virtual environment...
installing poetry...
Fatal error from pip prevented installation. Full pip output in file:
    /opt/pipx/logs/cmd_2024-07-22_16.55.46_pip_errors.log

Some possibly relevant errors from pip install:
    TypeError: Expected maxsize to be an integer or None
Error installing poetry.
```

multani replied:
Hm, I can't retry, but I don't think it's related to your change :)

multani commented on samples/openapi3/client/petstore/python-pydantic-v1/tests/test_api_client.py:
```diff
@@ -9,8 +9,6 @@
 $ pytest
 """
 
-import os
-import time
```
Can you revert this change? It has nothing to do with your original PR.

m-standfuss replied:
yep yep

multani reviewed (changes requested):
One last cleanup to remove, and it looks good to me afterwards :+1:

multani reviewed (approved):
Looks good to me :+1: 
Thanks for the change and the tests!
